### PR TITLE
swap

### PR DESCRIPTION
**Pull request by avp1598**

### CODE REVIEW

File: lib/types.ts [hunk 1]
```diff
@@ -369,3 +330,13 @@ export type MarketCapToTickParams = {
   totalSupply: bigint
   fee: 10000 | 3000 | 500 | 100
 }
+
+export interface GetTokenPoolResponse {
+  poolType: TokenPoolType
+  poolFee: number
+}
+
+export interface EstimateSwapResult {
```
Comment (hishboy) [inline lib/types.ts:339]: `SwapEstimation`

File: lib/constants.ts [hunk 2]
```diff
@@ -67,6 +74,10 @@ export const MULTISIG_FEE_COLLECTOR = EthAddressSchema.parse(
   '0x9A142B38d483d150dB2c115b4efA5ca37aC57Ebc'
 )
 
+export const BONDING_SWAP_CONTRACT = EthAddressSchema.parse(
+  '0x0000000000000000000000000000000000000000'
```
Comment (hishboy) [inline lib/constants.ts:78]: Did you test the smart contract with the SDK? you should deploy it and see if the SDK works with it

File: lib/Memecoin.ts [hunk 3]
```diff
@@ -1310,6 +1296,69 @@ export class MemecoinSDK {
     return BigInt(result.toFixed(0))
   }
 
+  private determinePoolOfMemecoin(memecoin: HydratedCoin): GetTokenPoolResponse {
```
Comment (hishboy) [inline lib/Memecoin.ts:1299]: resolveMemecoinPool

File: lib/Memecoin.ts [hunk 4]
```diff
@@ -1310,6 +1296,69 @@ export class MemecoinSDK {
...
+    if (!memecoin.dexInitiated) {
+      return { poolType: TokenPoolType.BondingCurve, poolFee: 0 }
+    }
+
+    switch (memecoin.dexKind) {
+      case 'univ2':
+        return { poolType: TokenPoolType.UniswapV2, poolFee: 0 }
+      case 'univ3':
+      case 'univ3-bonding':
+        return { poolType: TokenPoolType.UniswapV3, poolFee: 10000 }
```
Comment (hishboy) [inline lib/Memecoin.ts:1309]: don't hardcode the poolFee. fetch it

File: lib/Memecoin.ts [hunk 5]
```diff
@@ -466,39 +481,67 @@ export class MemecoinSDK {
...
+    const feeIn = tokenInData.poolFee
+    const tokenOutPoolType = tokenOutData.poolType
+    const feeOut = tokenOutData.poolFee
 
-    if (!isValidBigIntString(result)) {
-      throw new Error('Invalid response format')
+    const walletClient = this.walletClient
+    const account = walletClient.account
+    if (isNull(account)) {
+      throw new Error('No account found')
```
Comment (hishboy) [inline lib/Memecoin.ts:505]: wait, why do we need this? You're making the estimate function require a private key or account. We can estimate without a private key. we can pass it

File: lib/constants.ts (hunk 2 above)
Comment (hishboy) [reply lib/constants.ts:78]: also rename to SWAPPER_CONTRACT

File: lib/Memecoin.ts [hunk 6]
```diff
@@ -466,39 +481,67 @@ export class MemecoinSDK {
     return BigInt(result)
   }
 
-  private async estimateSwapCoin(params: EstimateSwapCoinParams): Promise<bigint> {
-    const { fromToken, toToken, amountIn } = params
+  async estimateSwap(params: EstimateSwapParams): Promise<EstimateSwapResult> {
```
Comment (hishboy) [inline lib/Memecoin.ts:484]: create a class called `TokenSwapper`, create an instance of it in the constructor. Ultimately we should see 

```typescript
return this.swapper.estimateSwap(params)
```

File: lib/Memecoin.ts [hunk 7]
```diff
@@ -919,21 +966,31 @@ export class MemecoinSDK {
...
   }
 
-  private async swapCoinFrontend(params: SwapFrontendParams): Promise<HexString> {
-    const { fromToken, toToken, amountIn, amountOut, slippage, affiliate } = params
-
-    const walletClient = this.walletClient
-
-    if (fromToken === 'eth' || toToken === 'eth') {
-      throw new Error('ETH is not supported as a fromToken or toToken')
+  async swap(params: SwapParams): Promise<HexString> {
```
Comment (hishboy) [inline lib/Memecoin.ts:969]: this method should looke like 

return this.swapper.swap(params)

File: lib/Memecoin.ts [hunk 8]
```diff
@@ -919,21 +966,31 @@ export class MemecoinSDK {
...
+      orderReferrer,
+      recipient,
+      feeIn,
+      feeOut,
+      tokenInPoolType,
+      tokenOutPoolType,
+      tokenOut
+    } = params
+
+    const amountOutMin = this.calculateMinAmountWithSlippage(amountOutMinimum)
```
Comment (hishboy) [inline lib/Memecoin.ts:990]: make `calculateMinAmountWithSlippage` a global function in `functions.ts` so you can use it here and in the swapper.ts code

File: lib/Memecoin.ts [hunk 9]
```diff
@@ -973,7 +1037,7 @@ export class MemecoinSDK {
       const result = await writeContracts(walletClient, {
         contracts: [
           // approve
-          ...(allowance < amountIn ? [approveContractCall] : []),
+          ...(tokenInPoolType !== TokenPoolType.WETH ? [approveContractCall] : []),
```
Comment (hishboy) [inline lib/Memecoin.ts:1040]: you still need to check allowance and prevent an extra approveContractCall call. You can do it during estimation phase and pass in the param

File: lib/Memecoin.ts [hunk 10]
```diff
@@ -1007,10 +1071,10 @@ export class MemecoinSDK {
 
       return lastReceipt.transactionHash
     } else {
-      if (allowance < amountIn) {
+      if (tokenInPoolType !== TokenPoolType.WETH) {
```
Comment (hishboy) [inline lib/Memecoin.ts:1074]: same

File: lib/Memecoin.ts [hunk 11]
```diff
@@ -466,39 +481,67 @@ export class MemecoinSDK {
...
-        toToken,
-        amountIn: amountIn.toString()
-      })
-    })
+    const tokenInData = tokenInMemecoin
+      ? this.determinePoolOfMemecoin(tokenInMemecoin)
+      : await this.findTokenWETHPool(params.tokenIn)
+    const tokenOutData = tokenOutMemecoin
+      ? this.determinePoolOfMemecoin(tokenOutMemecoin)
+      : await this.findTokenWETHPool(params.tokenOut)
```
Comment (hishboy) [inline lib/Memecoin.ts:495]: add an LRU cache for at least 1 min or something. we're going to call estimateSwap on every keyboard keystroke

cache value should be a promise so prevent multiple calls to server i.e. `const poolCache = LRUCache(EthAddress, Promise<GetTokenPoolResponse>`

File: lib/Memecoin.ts [hunk 12]
```diff
@@ -1310,6 +1296,69 @@ export class MemecoinSDK {
...
+
+      if (v2Response !== ZERO_ADDRESS) {
+        return {
+          poolType: TokenPoolType.UniswapV2,
+          poolFee: 0
+        }
+      } else {
+        throw new Error('No pool found')
+      }
+    }
```
Comment (hishboy) [inline lib/Memecoin.ts:1350]: can this as well in parallel. we're already caching things so it's not too bad

File: lib/Memecoin.ts [hunk 13]
```diff
@@ -1455,4 +1485,48 @@ export class MemecoinSDK {
         return UNISWAP_V3_ROUTER
     }
   }
+
+  private async getPoolLiquidityAtCurrentTick(poolAddress: EthAddress): Promise<bigint> {
```
Comment (hishboy) [inline lib/Memecoin.ts:1489]: create a UniswapV3 class under `uniswapv3.ts` and move this function to it. basically whoever needs this, will have to do 

```this.uniswapv3.getPoolLiquidityAtCurrentTick(poolAddress)```

File: lib/Memecoin.ts (hunk 4 above)
Comment (avp1598) [reply lib/Memecoin.ts:1309]: why do we fetch it? it is our memecoin right, we want them to go through our 10000 fee pool

**Pull request merged**